src/main/kotlin/org/eclipse/uprotocol/communication/InMemorySubscriber.kt

```diff
@@ -0,0 +1,210 @@
+    }
+
+    init {
+        scope.launch {
+            notifier.registerNotificationListener(NOTIFICATION_TOPIC, mNotificationListener)
+        }
+    }
+
+    /**
+     * Subscribe to a given topic.
+     *
+     * The API will return a [Result] with the response [SubscriptionResponse] or exception
+     * with the failure if the subscription was not successful. The API will also register the listener to be
+     * called when messages are received.
+     *
+     * @param topic The topic to subscribe to.
+     * @param listener The listener to be called when a message is received on the topic.
+     * @param options The call options for the subscription.
+     * @return Returns the [Result] with the response UMessage or exception with the failure
+     * reason as [UStatus].
+     */
+    override suspend fun subscribe(
+        topic: UUri,
+        listener: UListener,
+        options: CallOptions,
+        handler: SubscriptionChangeHandler?
+    ): Result<SubscriptionResponse> {
+
+        val request = subscriptionRequest {
+            this.topic = topic
+            this.subscriber = subscriberInfo { uri = transport.getSource() }
+        }
+
+        return rpcClient.invokeMethod(SUBSCRIBE_METHOD, UPayload.pack(request), options)
+            .mapToMessage<SubscriptionResponse>()
+            .onSuccess {
+                val state = it.status.state
+                if (state == SubscriptionStatus.State.SUBSCRIBED || state == SubscriptionStatus.State.SUBSCRIBE_PENDING) {
+                    transport.registerListener(topic, listener = listener)
+                }
+            }.mapCatching { response ->
+                handler?.let { handler ->
+                    mutex.withLock {
+                        if (mHandlers[topic] != null && mHandlers[topic] != handler) {
+                            throw UStatusException(UCode.ALREADY_EXISTS, "Handler already registered")
+                        } else if (mHandlers[topic] == handler) {
+                            response
+                        } else {
+                            mHandlers[topic] = handler
+                            response
+                        }
+                    }
+                } ?: response
+            }
+    }
+
+
+    /**
+     * Unsubscribe to a given topic.
+     *
+     * The subscriber no longer wishes to be subscribed to said topic so we issue a unsubscribe
+     * request to the USubscription service. The API will return a [UStatus]. If we are unable to
+     * unsubscribe to the topic with USubscription service, the listener and handler (if any) will remain registered.
+     *
+     * @param topic The topic to unsubscribe to.
+     * @param listener The listener to be called when a message is received on the topic.
+     * @param options The call options for the subscription.
+     * @return Returns [UStatus] with the result from the unsubscribe request.
+     */
+    override suspend fun unsubscribe(
```

Comment:
You're not removing the SubscriptionChangeHandler that might have been registered in the subscribe() API

Reply:
it is removed in line 135 if invokeMethod success